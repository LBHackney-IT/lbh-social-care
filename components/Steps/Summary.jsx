import { useState } from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';

import { Button } from 'components/Form';
import { deleteData } from 'utils/saveData';
import Summary from 'components/Summary/Summary';
import ErrorSummary from 'components/ErrorSummary/ErrorSummary';
import { filterDataOnCondition } from 'utils/steps';

const SummaryStep = ({ formData, formSteps, formPath, onFormSubmit }) => {
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [hasError, setHasError] = useState(false);
  const onSubmit = async () => {
    setIsSubmitting(true);
    setHasError(false);
    try {
      const data = await onFormSubmit(
        filterDataOnCondition(formSteps, formData)
      );
      deleteData(formPath);
      Router.replace(
        `${formPath}confirmation${data?.ref ? `?ref=${data.ref}` : ''}`
      );
    } catch {
      setHasError(true);
    }
    setIsSubmitting(false);
  };
  if (!formSteps) return null;
  return (
    <div>
      <Summary
        formData={formData}
        formPath={formPath}
        formSteps={formSteps}
        canEdit
      />
      <Button
        className="govuk-button"
        label="Submit"
        onClick={onSubmit}
        disabled={isSubmitting}
      />
      {hasError && (
        <ErrorSummary
          title="Unfortunately there was a problem with your submission."
          body="Please try again."
        />
      )}
    </div>
  );
};

SummaryStep.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  formSteps: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  formPath: PropTypes.string.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
};

export default SummaryStep;
