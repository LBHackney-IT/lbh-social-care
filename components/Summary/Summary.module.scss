.sectionTitle {
  display: flex;
  align-items: center;
  a {
    margin-left: 1rem;
  }
}

.summaryControls {
  display: flex;
  justify-content: flex-end;
}

.formattedTextArea {
  white-space: pre-line;
}
