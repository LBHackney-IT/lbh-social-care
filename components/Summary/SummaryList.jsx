import PropTypes from 'prop-types';

const SummaryList = ({ list }) => (
  <dl className="govuk-summary-list">
    {list &&
      list.map(({ key, title, value, href }) => (
        <div key={key} className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">{title}</dt>
          <dd className="govuk-summary-list__value">{value}</dd>
          {href && (
            <dd className="govuk-summary-list__actions">
              <a className="govuk-link" href="#">
                Change
                <span className="govuk-visually-hidden"> {title}</span>
              </a>
            </dd>
          )}
        </div>
      ))}
  </dl>
);

SummaryList.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.node.isRequired,
      href: PropTypes.string,
    }).isRequired
  ),
};

export default SummaryList;
