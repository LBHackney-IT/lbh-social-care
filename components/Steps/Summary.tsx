import React, { useState } from 'react';
import Router from 'next/router';

import Button from 'components/Button/Button';
import { deleteData } from 'utils/saveData';
import Summary from 'components/Summary/Summary';
import ErrorSummary from 'components/ErrorSummary/ErrorSummary';
import { filterDataOnCondition } from 'utils/steps';
import { sanitiseObject } from 'utils/objects';
import { FormStep } from 'components/Form/types';

interface Props {
  formData: Record<string, unknown>;
  formSteps: FormStep[];
  formPath: string;
  onFormSubmit: (filteredData: unknown) => Promise<{ ref?: string } | void>;
  isSummaryCollapsable?: boolean;
}

const SummaryStep = ({
  formData,
  formSteps,
  formPath,
  onFormSubmit,
  isSummaryCollapsable,
}: Props): React.ReactElement => {
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [hasError, setHasError] = useState(false);
  const onSubmit = async () => {
    setIsSubmitting(true);
    setHasError(false);
    try {
      const data = await onFormSubmit(
        //@ts-ignore TODO fix me
        filterDataOnCondition(formSteps, sanitiseObject(formData))
      );
      deleteData(formPath);
      Router.replace(
        `${formPath}confirmation${data && data.ref ? `?ref=${data.ref}` : ''}`
      );
    } catch {
      setHasError(true);
    }
    setIsSubmitting(false);
  };
  return (
    <div>
      <h2 className="lbh-heading-h1 govuk-!-margin-bottom-7">
        {`Review details`}
      </h2>
      <Summary
        formData={formData}
        formPath={formPath}
        formSteps={formSteps}
        canEdit
        isSummaryCollapsable={isSummaryCollapsable}
      />
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Button
          wideButton
          className="govuk-button"
          label="Cancel"
          isSecondary
          onClick={() => (window.location.href = '/')}
        />
        <Button
          wideButton
          className="govuk-button"
          label="Submit"
          onClick={onSubmit}
          disabled={isSubmitting}
        />
      </div>
      {hasError && (
        <ErrorSummary
          title="There was a problem with your submission."
          body="Please try again."
        />
      )}
    </div>
  );
};

export default SummaryStep;
