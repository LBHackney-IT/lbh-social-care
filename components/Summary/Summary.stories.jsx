import Summary from './Summary';

export default {
  title: 'Summary',
  component: Summary,
};

const Template = (args) => <Summary {...args} />;

export const Default = Template.bind({});
Default.args = {
  formData: {
    foo: '123',
    bar: 'asd',
    asd: '',
    qwe: 'yo',
    foobar: 'asd',
    textarea_verbose:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\nLorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    textarea_succinct: 'asd\nasd',
  },
  formPath: '/form/foo/',
  formSteps: [
    {
      id: 'first-step',
      title: 'First Step',
      components: [
        { component: 'TextInput', label: 'I am foo component', name: 'foo' },
        { component: 'TextInput', label: 'I am bar component', name: 'bar' },
        { component: 'TextInput', label: 'I am asd component', name: 'asd' },
      ],
    },
    {
      id: 'second-step',
      title: 'Second Step',
      components: [
        {
          component: 'TextInput',
          label: 'I am foobar component',
          name: 'foobar',
        },
      ],
    },
    {
      id: 'third-step',
      title: 'TextArea Step',
      components: [
        {
          component: 'TextArea',
          label: 'I am a verbose textarea',
          name: 'textarea_verbose',
        },
        {
          component: 'TextArea',
          label: 'I am a succinct textarea',
          name: 'textarea_succinct',
        },
      ],
    },
  ],
  onFormSubmit: () => alert('submit'),
};
