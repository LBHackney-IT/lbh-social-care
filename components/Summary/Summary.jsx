import { useState } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import Link from 'next/link';

import SummaryList from 'components/Summary/SummaryList';
import { filterStepsOnCondition, filterDataOnCondition } from 'utils/steps';
import { formatData, getSectionObject } from 'utils/summary';
import { setValues } from 'utils/objects';

import styles from './Summary.module.scss';

const SummaryMultiSection = ({
  formData,
  id,
  title,
  canEdit,
  formPath,
  ...props
}) => (
  <>
    {formData[id].map((data, key) => (
      <SummarySection
        {...props}
        canEdit={canEdit}
        formData={data}
        formPath={formPath}
        title={`${title} - ${key + 1}`}
        key={`${id}/${key}`}
        id={`${id}/${key + 1}`}
      />
    ))}
    {canEdit && (
      <p className="govuk-!-margin-bottom-7">
        <Link
          href={`${formPath}${id}/${formData[id].length + 1}?fromSummary=true`}
          as={`${formPath}${id}/${formData[id].length + 1}?fromSummary=true`}
        >
          <a className="govuk-link">Add Another {title}</a>
        </Link>
      </p>
    )}
  </>
);

SummaryMultiSection.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      type: PropTypes.string,
      href: PropTypes.string,
    })
  ),
};

export const SummarySection = ({
  formData,
  id,
  title,
  components,
  formPath,
  canEdit,
  collapsedSection,
  toggleCollapsed,
  isSummaryCollapsable,
  additionalMetadata,
}) => {
  const isCollapsed = collapsedSection[id];

  return (
    <div className="govuk-!-margin-bottom-7">
      <div className="lbh-table-header">
        <div className={styles.sectionTitle}>
          <h3 className="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0">
            {title.toUpperCase()}
          </h3>
          {canEdit && (
            <Link
              href={`${formPath}${id}?fromSummary=true`}
              as={`${formPath}${id}?fromSummary=true`}
            >
              <a className="govuk-link">Edit</a>
            </Link>
          )}
        </div>
        {isSummaryCollapsable && (
          <button
            className="govuk-link govuk-link--underline"
            onClick={(e) => {
              e.preventDefault();
              toggleCollapsed(id);
            }}
          >
            {isCollapsed ? 'Expand view' : 'Collapse view'}
          </button>
        )}
      </div>
      <hr className="govuk-divider" />
      {!isCollapsed && (
        <SummaryList
          list={[
            ...components
              .filter(({ name }) => formData[name])
              .map((data) => formatData(data, formData)),
            ...(additionalMetadata || []),
          ]}
        />
      )}
    </div>
  );
};

SummarySection.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  components: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
    }).isRequired
  ).isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
  collapsedSection: PropTypes.object.isRequired,
  toggleCollapsed: PropTypes.func.isRequired,
  isSummaryCollapsable: PropTypes.bool,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      type: PropTypes.string,
      href: PropTypes.string,
    })
  ),
};

const Summary = ({
  formData,
  formSteps,
  formPath,
  canEdit,
  isSummaryCollapsable,
  additionalMetadata,
}) => {
  const [collapsedSection, setCollapsedSection] = useState(
    getSectionObject(formSteps, formData, false)
  );
  const hasCollapsed = Object.values(collapsedSection).find(Boolean);
  return (
    <>
      {isSummaryCollapsable && (
        <div className={styles.summaryControls}>
          <button
            className={cx(
              'govuk-link',
              'govuk-link--underline',
              'govuk-!-font-weight-bold'
            )}
            onClick={(e) => {
              e.preventDefault();
              setCollapsedSection(setValues(collapsedSection, !hasCollapsed));
            }}
          >
            {hasCollapsed ? 'Expand all' : 'Collapse all'}
          </button>
        </div>
      )}
      {filterStepsOnCondition(formSteps, formData).map((section) => {
        const props = {
          key: section.id,
          formData: filterDataOnCondition(formSteps, formData),
          formPath: formPath,
          canEdit: canEdit,
          collapsedSection,
          toggleCollapsed: (id) =>
            setCollapsedSection({
              ...collapsedSection,
              [id]: !collapsedSection[id],
            }),
          isSummaryCollapsable,
          ...section,
          additionalMetadata,
        };
        return section.isMulti ? (
          <SummaryMultiSection {...props} />
        ) : (
          <SummarySection {...props} />
        );
      })}
    </>
  );
};

Summary.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  formSteps: PropTypes.array.isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
  isSummaryCollapsable: PropTypes.bool,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      type: PropTypes.string,
      href: PropTypes.string,
    })
  ),
};

export default Summary;
