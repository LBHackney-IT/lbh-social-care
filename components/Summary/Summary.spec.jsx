import { render, fireEvent } from '@testing-library/react';

import Summary from './Summary';

describe('Summary component', () => {
  const objectComponents = [
    {
      component: 'TextInput',
      name: 'first',
    },
    {
      component: 'TextInput',
      name: 'second',
    },
    {
      component: 'TextInput',
      name: 'third',
    },
  ];
  const props = {
    formData: {
      foo: '123',
      bar: 'asd',
      asd: '',
      qwe: 'yo',
      foobar: 'asd',
      oo: { first: 'first', second: 'second', third: 'third' },
      oi: { first: 'first', second: 'second', third: 'third' },
    },
    formPath: '/form/foo/',
    formSteps: [
      {
        id: 'first-step',
        title: 'First Step',
        components: [
          { component: 'TextInput', label: 'I am foo component', name: 'foo' },
          { component: 'TextInput', label: 'I am bar component', name: 'bar' },
          { component: 'TextInput', label: 'I am asd component', name: 'asd' },
        ],
      },
      {
        id: 'second-step',
        title: 'Second Step',
        components: [
          {
            component: 'TextInput',
            label: 'I am foobar component',
            name: 'foobar',
          },
        ],
      },
      {
        id: 'object-step',
        title: 'Object Step',
        components: [
          {
            component: 'ObjectInput',
            label: 'I am object component',
            name: 'oo',
            components: objectComponents,
          },
          {
            component: 'ObjectInput',
            label: 'I am object component inline',
            name: 'oi',
            summaryInline: true,
            components: objectComponents,
          },
        ],
      },
    ],
    onFormSubmit: jest.fn(),
  };

  it('should render properly', () => {
    const { asFragment } = render(<Summary {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('isSummaryCollapsable', () => {
    const localProps = {
      formData: {
        foo: '123',
      },
      formPath: '/form/foo/',
      formSteps: [
        {
          id: 'first-step',
          title: 'First Step',
          components: [
            {
              component: 'TextInput',
              label: 'I am foo component',
              name: 'foo',
            },
          ],
        },
      ],
    };

    it('should render properly - with isSummaryCollapsable', () => {
      const { asFragment, getByText, queryByText } = render(
        <Summary {...localProps} isSummaryCollapsable />
      );
      expect(asFragment()).toMatchSnapshot();
      expect(getByText('Collapse all')).toBeInTheDocument();
      expect(queryByText('I am foo component')).toBeInTheDocument();

      fireEvent.click(getByText('Collapse view'));
      expect(queryByText('Collapse all')).not.toBeInTheDocument();
      expect(queryByText('Expand all')).toBeInTheDocument();

      fireEvent.click(getByText('Expand all'));
      expect(queryByText('Collapse all')).toBeInTheDocument();
      expect(queryByText('Expand all')).not.toBeInTheDocument();
      expect(queryByText('I am foo component')).toBeInTheDocument();
    });

    it('should render properly - with No isSummaryCollapsable', () => {
      const { asFragment } = render(<Summary {...localProps} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should render any additional metadata provided', () => {
    const { getByText } = render(
      <Summary
        formData={{
          foo: '123',
        }}
        formPath="/form/foo/"
        formSteps={[
          {
            id: 'first-step',
            title: 'First Step',
            components: [
              {
                component: 'TextInput',
                label: 'I am foo component',
                name: 'foo',
              },
            ],
          },
        ]}
        additionalMetadata={[
          {
            key: 'title-of-row',
            title: 'Title of row',
            value: 'Value of row',
          },
        ]}
      />
    );

    getByText('Title of row');
    getByText('Value of row');
  });
});
