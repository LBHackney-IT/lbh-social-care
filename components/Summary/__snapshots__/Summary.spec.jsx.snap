// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary component isSummaryCollapsable should render properly - with No isSummaryCollapsable 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <div
        class="sectionTitle"
      >
        <h3
          class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
        >
          FIRST STEP
        </h3>
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am foo component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          123
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;

exports[`Summary component isSummaryCollapsable should render properly - with isSummaryCollapsable 1`] = `
<DocumentFragment>
  <div
    class="summaryControls"
  >
    <button
      class="govuk-link govuk-link--underline govuk-!-font-weight-bold"
    >
      Collapse all
    </button>
  </div>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <div
        class="sectionTitle"
      >
        <h3
          class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
        >
          FIRST STEP
        </h3>
      </div>
      <button
        class="govuk-link govuk-link--underline"
      >
        Collapse view
      </button>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am foo component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          123
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;

exports[`Summary component should render properly 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <div
        class="sectionTitle"
      >
        <h3
          class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
        >
          FIRST STEP
        </h3>
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am foo component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          123
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am bar component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          asd
        </dd>
      </div>
    </dl>
  </div>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <div
        class="sectionTitle"
      >
        <h3
          class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
        >
          SECOND STEP
        </h3>
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am foobar component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          asd
        </dd>
      </div>
    </dl>
  </div>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <div
        class="sectionTitle"
      >
        <h3
          class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
        >
          OBJECT STEP
        </h3>
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am object component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <div>
            <span>
              first
            </span>
            <br />
          </div>
          <div>
            <span>
              second
            </span>
            <br />
          </div>
          <div>
            <span>
              third
            </span>
            <br />
          </div>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am object component inline
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <span>
            first 
          </span>
          <span>
            second 
          </span>
          <span>
            third 
          </span>
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
